### ISSUE TITLE
SR.py

### ISSUE THREAD

**Issue by XishunZhu** [XishunZhu/Garesat_Net-SR-#3]
# -*- coding:utf-8 -*-
# author: Xishun Zhu
# datetime: 2022/5/27 18:16
import argparse
import os
import shutil
import socket
import time
import numpy as np
import PIL.Image as Image

import torch
import torch.backends.cudnn as cudnn
import torch.nn as nn
import torch.nn.parallel
import torch.optim as optim
import torch.utils.data
import torchvision.utils as vutils
import torchvision.transforms as tt

from torch.autograd import Variable
from torch.optim.lr_scheduler import ReduceLROnPlateau
from torch.utils.data import DataLoader
from torch.utils.tensorboard import SummaryWriter
from Mytextureloss import LosstextureModel
from models.Garesat import RResNet
from mydct import Mymatrix, myresize
import utils.transformed as transforms
from data.ImageFolderDataset import MyImageFolder

DATA_DIR = 'E:\Xishun\supercostruction\data'

parser = argparse.ArgumentParser ()
parser.add_argument ('--dataset', default="train",
                     help='train | val | test')
parser.add_argument ('--workers', type=int, default=1,
                     help='number of data loading workers')
parser.add_argument ('--batchSize', type=int, default=1,
                     help='input batch size')
parser.add_argument ('--resume', default='True',
                     help='load model resume')
parser.add_argument ('--niter', type=int, default=10000,
                     help='number of epochs to train for')
parser.add_argument ('--lr', type=float, default=0.001,
                     help='learning rate, default=0.001')
parser.add_argument ('--decay_round', type=int, default=10,
                     help='learning rate decay 0.9 each decay_round')
parser.add_argument ('--beta1', type=float, default=0.9,
                     help='beta1 for adam. default=0.9')
parser.add_argument ('--cuda', type=bool, default=True,
                     help='enables cuda')
parser.add_argument ('--ngpu', type=int, default=1,
                     help='number of GPUs to use')
parser.add_argument ('--Re', default='',
                     help="path to PSResNet1 (to continue training)")
parser.add_argument ('--trainpics', default='./training/',
                     help='folder to output training images')
parser.add_argument ('--validationpics', default='./training/',
                     help='folder to output validation images')
parser.add_argument ('--testPics', default='./training/',
                     help='folder to output test images')
parser.add_argument ('--outckpts', default='./training/',
                     help='folder to output checkpoints')
parser.add_argument ('--outlogs', default='./training/',
                     help='folder to output images')
parser.add_argument ('--outcodes', default='./training/',
                     help='folder to save the experiment codes')
parser.add_argument ('--beta', type=float, default=1.8,
                     help='hyper parameter of beta')
# parser.add_argument('--gpu_ids', type=str, default='0,1')
parser.add_argument ('--start_epoch', default=0,
                     help='load model start_epoch')

parser.add_argument ('--remark', default='', help='comment')

parser.add_argument ('--test', default='', help='test mode, you need give the test pics dirs in this param')

parser.add_argument ('--hostname', default=socket.gethostname (), help='the  host name of the running server')
parser.add_argument ('--debug', type=bool, default=False, help='debug mode do not create folders')
parser.add_argument ('--logFrequency', type=int, default=10, help='the frequency of print the log on the console')
parser.add_argument ('--resultPicFrequency', type=int, default=1, help='the frequency of save the resultPic')

os.environ["CUDA_VISIBLE_DEVICES"] = '0'

def weights_init(m):
    classname = m.__class__.__name__
    if classname.find ('Conv') != -1:
        m.weight.data.normal_ (0.0, 0.02)
    elif classname.find ('BatchNorm') != -1:
        m.weight.data.normal_ (1.0, 0.02)
        m.bias.data.fill_ (0)

# print the structure and parameters number of the net
def print_network(net):
    num_params = 0
    for param in net.parameters ():
        num_params += param.numel ()
    print_log (str (net), logPath)
    print_log ('Total number of parameters: %d' % num_params, logPath)

# save code of current experiment
def save_current_codes(des_path):
    main_file_path = os.path.realpath (__file__)

    cur_work_dir, mainfile = os.path.split (main_file_path)

    new_main_path = os.path.join (des_path, mainfile)
    shutil.copyfile (main_file_path, new_main_path)
    data_dir = cur_work_dir + "/data/"
    new_data_dir_path = des_path + "/data/"
    shutil.copytree(data_dir, new_data_dir_path)

    model_dir = cur_work_dir + "/models/"
    new_model_dir_path = des_path + "/models/"
    shutil.copytree (model_dir, new_model_dir_path)

    utils_dir = cur_work_dir + "/utils/"
    new_utils_dir_path = des_path + "/utils/"
    shutil.copytree (utils_dir, new_utils_dir_path)

    # return errR1

def main():
    ############### define global parameters ###############
    global opt, optimizerR,  writer, batchSize, logPath,  schedulerR, val_loader, smallestLoss, val_dataset, train_dataset, test_dataset, start_epoch

    #################  output configuration   ###############
    opt = parser.parse_args ()  # 
    
    if torch.cuda.is_available () and not opt.cuda:
        print ("WARNING: You have a CUDA device, "
               "so you should probably run with --cuda")

    cudnn.benchmark = True

    ############  create dirs to save the result #############
    if not opt.debug:
        try:
            cur_time = time.strftime ('%Y-%m-%d-%H_%M_%S', time.localtime ())
            experiment_dir = opt.hostname + "_" + cur_time + opt.remark
            # experiment_dir = opt.hostname + "_" + opt.remark
            opt.outckpts += experiment_dir + "/checkPoints"
            opt.trainpics += experiment_dir + "/trainPics"
            opt.validationpics += experiment_dir + "/validationPics"
            opt.outlogs += experiment_dir + "/trainingLogs"
            opt.outcodes += experiment_dir + "/codes"
            opt.testPics += experiment_dir + "/testPics"
            if not os.path.exists (opt.outckpts):
                # shutil.rmtree(opt.outckpts)
                os.makedirs (opt.outckpts)
            if not os.path.exists (opt.trainpics):
                # shutil.rmtree(opt.trainpics)
                os.makedirs (opt.trainpics)
            if not os.path.exists (opt.validationpics):
                # shutil.rmtree(opt.validationpics)
                os.makedirs (opt.validationpics)
            if not os.path.exists (opt.outlogs):
                # shutil.rmtree(opt.outlogs)
                os.makedirs (opt.outlogs)
            if not os.path.exists (opt.outcodes):
                # shutil.rmtree(opt.outcodes)
                os.makedirs (opt.outcodes)
            # else:
            #     shutil.rmtree(opt.outcodes)
            if (not os.path.exists (opt.testPics)) and opt.test != '':
                # shutil.rmtree(opt.testPics)
                os.makedirs (opt.testPics)

        except OSError:
            print ("mkdir failed   XXXXXXXXXXXXXXXXXXXXX")

    logPath = opt.outlogs + '/%s_%d_log.txt' % (opt.dataset, opt.batchSize)  # 

    print_log (str (opt), logPath)
    if os.path.exists (opt.outcodes):
        shutil.rmtree (opt.outcodes)
    else:
        os.makedirs (opt.outcodes)
        save_current_codes (opt.outcodes)

    if opt.test == '':
        # tensorboardX writer
        # writer = SummaryWriter(comment='**' + opt.remark)
        writer = SummaryWriter (comment='_' + opt.remark)
        ##############   get dataset   ############################得
        traindir = os.path.join (DATA_DIR, 'trainPics')  #
        valdir = os.path.join(DATA_DIR, 'val')          #

        train_dataset = MyImageFolder (
            traindir,
            transforms.Compose ([
                transforms.ToTensor (),
            ]))
        val_dataset = MyImageFolder(
            valdir,
            transforms.Compose([
                transforms.ToTensor(),
            ]))

        assert train_dataset
        assert val_dataset

    batch_size = opt.batchSize

    Re = Recover ()
    Re.cuda ()
    Re.apply (weights_init)  # apply
    if opt.Re != '':
        Re.load_state_dict (torch.load (opt.Re1))
    if opt.ngpu > 1:
        Re = torch.nn.DataParallel (Re).cuda ()
    print_network (Re)

    Losstexture=LosstextureModel()
    criterion = nn.MSELoss ()

    # training mode
    if opt.test == '':
        optimizerR = optim.Adam(Re.parameters(), lr=opt.lr, betas=(opt.beta1, 0.9))
        schedulerR = ReduceLROnPlateau(optimizerR, mode='min', factor=0.98, patience=20, verbose=True)

        train_loader = DataLoader(train_dataset, batch_size=opt.batchSize,
                                   num_workers=int (opt.workers), drop_last=True)

        val_loader = DataLoader(val_dataset, batch_size=opt.batchSize,
                                shuffle=True, num_workers=int(opt.workers),drop_last=True)
        smallestLoss = 10000
        print_log ("training is beginning .......................................................", logPath)
        # ---------------------------------- Breakpoint continuation training——————————————————————————--------
        if opt.resume:
            print ('yes')
            if os.path.isfile (
                    r'E:\Xishun\supercostruction\training\DESKTOP-M9QJC45_2022-06-19-09_43_55\checkPoints/Rcheckpoint_model_epoch_1240pth'):
                checkpoint = torch.load (
                    r'E:\Xishun\supercostruction\training\DESKTOP-M9QJC45_2022-06-19-09_43_55\checkPoints/Rcheckpoint_model_epoch_1240pth')
                Re.load_state_dict (checkpoint['model'], False)
                optimizerR.load_state_dict (checkpoint['optimizer'])
                start_epoch = checkpoint['epoch'] + 1
                print ('Load epoch {} successly！'.format (start_epoch))
            else:
                start_epoch = 0
                print ('No saved model, training will start from scratch！')

        for epoch in range (start_epoch, opt.niter):
        # for epoch in range (0, opt.niter):
            ######################## train ##########################################
            print (" ######################## train ##########################################")
            
            errR1 = train (train_loader, epoch, Re=Re, criterion=criterion,Losstexture=Losstexture)          
            schedulerR.step (errR1)
    
            if epoch % 20 == 0 :
                # print(" ######################## validation ##########################################")
                validation(val_loader, epoch,  Re=Re ,criterion=criterion,Losstexture=Losstexture)

                Rfilepath = os.path.join(opt.outckpts, 'Rcheckpoint_model_epoch_{}.pth'.format(epoch))
                Rfilepath = Rfilepath.replace('.', '')
                Rfilepath = Rfilepath.replace('\\', '/')
                Rfilepath = r'E:\Xishun\supercostruction/' + R1filepath
              
              
              
                # -----------------------save epoch model---------------------
                Rstate = {'model': Re.state_dict(), 'optimizer': optimizerR.state_dict(), 'epoch': epoch}
                torch.save(Rstate, Rfilepath)
                print(Rfilepath)
              
            

        writer.close ()

def train(train_loader, epoch, Re,criterion,Losstexture):

    global this_batch_size, R1losses
    batch_time = AverageMeter ()
    data_time = AverageMeter ()
    this_batch_size = opt.batchSize

    R1losses = AverageMeter()
    R2losses = AverageMeter()
    # R3losses = AverageMeter()

    Re.train()
    start_time = time.time ()

    for i, [label2, label1,image] in enumerate (train_loader, 0):

        data_time.update (time.time () - start_time)
        # _, _, h, w = image.shape
        # image1 = copy.copy(image)
        # image1 = np.array(image1)
        # image1 = myresize.Myresize(image1, h/2,w/2)
        Re.zero_grad ()

        if opt.cuda:
            image = image.cuda()
            label1 = label1.cuda()
            label2 = label2.cuda()

        image = Variable (image)
        label1 = Variable (label1)
        label2 = Variable (label2)

        reimage1 = Re(image)
        errR_11 = criterion (reimage1, label1)
        errR_12= Losstexture(reimage1, label1)
        errR1=(errR_11+errR_12)/2
        R1losses.update (errR1.item (), this_batch_size)
      
        reimage1 = reimage1.detach()
        reimage2 = Re(reimage1)
        errR_21 = criterion(reimage2, label2)
        errR_22 = Losstexture(reimage2, label2)
        errR2=(errR_21+errR_22)/2
        R2losses.update (errR2.item (), this_batch_size)
       

        # reimage2 = reimage2.detach()
        # reimage3 = Re(reimage2)
        # errR_31 = criterion(reimage3, label2)
        # errR_32 = Losstexture(reimage3, label2)
        errR3=(errR_31+errR_32)/2
        R3losses.update(errR3.item(), this_batch_size)
       

        R1losses.backward ()
        R2losses.backward()

        optimizerR.step()

        batch_time.update (time.time () - start_time)
        start_time = time.time ()

        log = '[%d/%d][%d/%d]/t Loss_R1:%.4f  Loss_R2:%.4f /tdatatime: %.4f/tbatchtime: %.4f' % (
            epoch, opt.niter, i, len (train_loader),
            errR1,errR2, data_time.val, batch_time.val)

        if i % opt.logFrequency == 0:
            print_log(log, logPath)
        else:
            print_log(log, logPath, console=False)

        if epoch % 20 == 0 and i == opt.resultPicFrequency  and epoch >0:

            save_result_pic(this_batch_size,  Rimage1,Rimage2,label1,label2,epoch, i,
                            opt.trainpics)

    # epcoh log
    epoch_log = "one epoch time is %.4f======================================================================" % (
        batch_time.sum) + "\n"
    epoch_log = epoch_log + "epoch learning rate:      optimizerR_lr = %.8f" % (
        optimizerR.param_groups[0]['lr']) + "\n"
    epoch_log = epoch_log + "epoch_R1loss=%.6f epoch_R2loss=%.6f\t" % (
         R1losses.avg, R1losses.avg)
    print_log(epoch_log, logPath)

    if not opt.debug:
        # record lr

        writer.add_scalar("lr/R_lr", optimizerR.param_groups[0]['lr'], epoch)

        writer.add_scalar("lr/beta", opt.beta, epoch)

        writer.add_scalar('train/R1_loss', R1losses.avg, epoch)

        writer.add_scalar('train/R_2_loss', R2losses.avg, epoch)
        # writer.add_scalar('train/R_3_loss', R3losses.avg, epoch)
    return errR1

def validation(val_loader, epoch, Re,criterion,Losstexture):
    global all_pics, this_batch_size, img_key, Bz, width, height, deep, errH, errR1, errR2, errR3, errR4, errR5, errR6,errR7,valcover_img, valsecret_img
    batch_time = AverageMeter()
    data_time = AverageMeter()

    valR1losses = AverageMeter()
    valR2losses = AverageMeter()
    valR3losses = AverageMeter()

    Re.eval()

    start_time = time.time()

    for i, [label2, label1, image] in enumerate(val_loader, 0):

        data_time.update(time.time() - start_time)
        # _, _, h, w = image.shape
        # image1 = copy.copy(image)
        # image1 = np.array(image1)
        # image1 = myresize.Myresize(image1, h/2,w/2)
        Re.zero_grad()
        if opt.cuda:
            image = image.cuda()
            label1 = label1.cuda()
            label2 = label2.cuda()

        image = Variable(image)
        label1 = Variable(label1)
        label2 = Variable(label2)

        reimage1 = Re(image)
        errR11 = criterion(reimage1, label1)
        errR12 = Losstexture(reimage1, label1)
        errR1 =(errR11+errR12)/2
        valR1losses.update(errR1.item(), this_batch_size)
      
        reimage1 = reimage1.detach()
        reimage2 = Re(reimage1)
        errR21 = criterion(reimage2, label2)
        errR22 = Losstexture(reimage2, label2)
        errR2 =(errR21+errR22)/2
        valR2losses.update(errR2.item(), this_batch_size)
       
        # reimage2 = reimage2.detach()
        # reimage3 = Re(reimage2)
        # errR31 = criterion(reimage3, label2)
        # errR33 = Losstexture(reimage3, label2)
        #errR3 =(errR31+errR32)/2
        #valR3losses.update(errR3.item(), this_batch_size)
       
        batch_time.update(time.time() - start_time)
        start_time = time.time()

        log = '[%d/%d][%d/%d]/t Loss_R1: %.4f Loss_R2: %.4f/tdatatime: %.4f/tbatchtime: %.4f' % (
            epoch, opt.niter, i, len(val_loader),
             valR1losses.val,valR2losses.val, data_time.val, batch_time.val)

        if i % opt.logFrequency == 0:
            print_log(log, logPath)
        else:
            print_log(log, logPath, console=False)

        if epoch % 20 == 0 and i == opt.resultPicFrequency and epoch > 0:
            save_result_pic(this_batch_size, Rimage1, Rimage2, label1, label2, epoch, i,
                            opt.trainpics)
    # epcoh log
    epoch_log = "one epoch time is %.4f======================================================================" % (
        batch_time.sum) + "\n"
    epoch_log = epoch_log + "\tepoch_R1Loss=%.6f epoch_R2Loss=%.6f " % (
         valR1losses.avg,valR2losses.avg,)
    print_log(epoch_log, logPath)

    if not opt.debug:
        # record loss
        writer.add_scalar('val/R_loss', valR1losses.avg, epoch)

        writer.add_scalar('val/R_2_loss', valR2losses.avg, epoch)
        writer.add_scalar('val/R_3_loss', valR3losses.avg, epoch)

def print_log(log_info, log_path, console=True):
    # print info onto the console
    if console:
        print (log_info)
    # debug mode will not write logs into files
    if not opt.debug:
        # write logs into log file
        if not os.path.exists (log_path):
            fp = open (log_path, "w")
            fp.writelines (log_info + "\n")
        else:
            with open (log_path, 'a+') as f:
                f.writelines (log_info + '\n')

def save_result_pic(this_batch_size, reimage1, reimage2, label1, label2, epoch, i, save_path):
    if not opt.debug:

        resultRevealImgName = '%s/reimage1_epoch%03d_batch%04d.png' % (save_path, epoch, i)
        vutils.save_image (reimage1, resultRevealImgName, nrow=this_batch_size, padding=1, normalize=True)

        resultrevsecret16 = '%s/reimage2_epoch%03d_batch%04d.png' % (save_path, epoch, i)
        vutils.save_image (reimage2, resultrevsecret16, nrow=this_batch_size, padding=1, normalize=True)
        resultRevealImgName1 = '%s/label1_epoch%03d_batch%04d.png' % (save_path, epoch, i)
        vutils.save_image (label1, resultRevealImgName1, nrow=this_batch_size, padding=1, normalize=True)

        resultrevsecret162 = '%s/label1_epoch%03d_batch%04d.png' % (save_path, epoch, i)
        vutils.save_image (label2, resultrevsecret162, nrow=this_batch_size, padding=1, normalize=True)

class AverageMeter (object):
    """
    Computes and stores the average and current value.
    """

    def __init__(self):
        self.reset ()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count

if __name__ == '__main__':
    main ()
    pin_memory = True
